src/crafting.cpp

```diff
@@ -561,7 +554,7 @@ const inventory &Character::crafting_inventory( const tripoint &src_pos, int rad
         cached_crafting_inventory->form_from_map( inv_pos, radius, this, false, clear_path );
     }
 
-    for( const item_location &it : all_items_loc() ) {
+    for( const item_location &it : const_cast<Character *>( this )->all_items_loc() ) {
```

Comment:
So the morally correct thing here would be to introduce a new `const_item_location` class which is like `item_location`, except with const access to the underlying item, and then have a `const` overload of `all_items_loc` which uses that.  It looks like `cached_crafting_inventory.add_item` copies its argument, so I think that should work.

That's the sort of thing that could reasonably go in a follow-up PR.

Reply:
I still have a bit of PTSD from the last time I tried propagating const-ness to item_location. I would definitely prefer to leave it for another PR since even if I succeed with this new approach it will introduce a lot of extra diffs.

Reply:
Can you add a comment suggesting that this would be a good thing to add, perhaps?

Reply:
Done.